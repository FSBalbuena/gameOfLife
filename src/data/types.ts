export type cell = 0 | 1;
export type Grid = cell[][];

export type BoardErr = null | {
  steps?: string;
  grid?: string;
};

export type RequestErr = null | {
  steps?: string[];
  grid?: string[];
};
