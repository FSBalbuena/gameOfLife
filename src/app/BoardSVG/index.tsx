import { useCallback, ReactEventHandler } from "react";

type Props = {
  onUpdate: (x: number, y: number) => void;
  grid: boolean[][];
  range: number;
};

export default function BoardSVG({ onUpdate, grid, range }: Props) {
  const handleClick: ReactEventHandler<SVGSVGElement> = useCallback(
    (event) => {
      const cell = event.target as SVGSVGElement & {
        ariaRowIndex: number;
        ariaColIndex: number;
      };
      const { ariaRowIndex: x, ariaColIndex: y } = cell;
      onUpdate(Number(x), Number(y));
    },
    [onUpdate]
  );

  return (
    <svg
      role="application"
      className="w-full sm:w-2/3 aspect-square border"
      onClick={handleClick}
    >
      {grid.map((row, rowIndex) =>
        row.map((cell, colIndex) => (
          <rect
            role="button"
            key={`${rowIndex}-${colIndex}`}
            x={`${(rowIndex * 100) / range}%`}
            y={`${(colIndex * 100) / range}%`}
            aria-rowindex={rowIndex}
            aria-colindex={colIndex}
            width={`${100 / range}%`}
            height={`${100 / range}%`}
            fill={cell ? "black" : "white"}
            className="stroke-gray-400 stroke-1"
          />
        ))
      )}
    </svg>
  );
}
