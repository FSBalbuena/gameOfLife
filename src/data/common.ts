import { Grid } from "@/data/types";
import { z } from "zod";

export const RANGE = 30;

export const EMPTY_GRID: Grid = Array.from({ length: RANGE }, () =>
  Array.from({ length: RANGE }, () => 0)
);
import {
  REQUIRED_GRID,
  REQUIRED_STEP,
  INVALID_STEP_INT,
  INVALID_GRID,
  INVALID_GRID_CELL,
  INVALID_STEP_MIN,
  INVALID_STEP_NUMBER,
} from "@/data/copy";

//---request schema
export const stepsSchema = z
  .number({
    required_error: REQUIRED_STEP,
    invalid_type_error: INVALID_STEP_NUMBER,
  })
  .int(INVALID_STEP_INT)
  .positive(INVALID_STEP_MIN)
  .optional();

export const requestSchema = z.object({
  grid: z.array(
    z.array(
      z.number({
        invalid_type_error: INVALID_GRID_CELL,
      })
    ),
    {
      required_error: REQUIRED_GRID,
      invalid_type_error: INVALID_GRID,
    }
  ),
  steps: stepsSchema,
});
