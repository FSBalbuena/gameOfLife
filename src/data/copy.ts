//---copy
export const SUCCESS = "success";
export const FAIL = "fail";
export const REQUIRED_GRID = "Board is required";
export const INVALID_GRID = "Board must be a 2D array of numbers";
export const INVALID_GRID_CELL = "Invalid Cell value";
export const REQUIRED_STEP = "Step is required";
export const INVALID_STEP_NUMBER = "Step must be a number";
export const INVALID_STEP_INT = "Step must be an integer";
export const INVALID_STEP_MIN = "Step must be a positive number";
export const STANDAR_ERR =
  "Sorry, something went wrong with next state calculation";

export const GAME_TITLE = "Conway's Game of Life";
export const GAME_DESCRIPTION = "A cellular automaton simulation.";
export const GAME_LOADING_DESCRIPTION = "Generating next state.";
export const GAME_NEXT_BUTTON = "Next state";
export const GAME_ADVANCE_BUTTON = "Advance state";
export const GAME_PLAY_FOREVER_CHECKBOX = "Play Forever";
