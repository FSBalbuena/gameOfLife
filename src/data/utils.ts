import { BoardErr, Grid, RequestErr } from "@/data/types";

export const createGridByRange = (range: number): Grid =>
  Array.from({ length: range }, () => Array.from({ length: range }, () => 0));

export const parseRequestErrToBoardErr = (error: RequestErr): BoardErr => ({
  steps: error?.steps?.[0],
  grid: error?.grid?.[0],
});
