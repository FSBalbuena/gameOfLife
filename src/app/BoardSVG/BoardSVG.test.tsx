import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { createGridByRange } from "@/data/utils";
import BoardSVG from "@/app/BoardSVG";

const range = 5; // 5x5 grid
const grid = createGridByRange(range);
const onUpdate = jest.fn();

const defaultProps = {
  range,
  grid,
  onUpdate,
};

describe("BoardSVG", () => {
  it("Renders without crashing", () => {
    const { container } = render(<BoardSVG {...defaultProps} />);

    // Take the snapshot after the state update
    expect(container).toMatchSnapshot();
  });

  it("Renders the correct number of cells", () => {
    render(<BoardSVG {...defaultProps} />);
    const cellGroup = screen.getAllByRole("button");
    expect(cellGroup.length).toBe(defaultProps.range ** 2);
  });

  it("Renders the correct state for cells", () => {
    const statedGrid = grid.map((row) => [...row]);
    statedGrid[0][3] = 1;
    render(<BoardSVG {...defaultProps} grid={statedGrid} />);
    const cellGroup = screen.getAllByRole("button");
    const liveCell = cellGroup.at(3);
    expect(liveCell).toHaveAttribute("fill", "black");
  });

  describe("When a cell is clicked", () => {
    it("onUpdate should be called", () => {
      render(<BoardSVG {...defaultProps} />);
      expect(defaultProps.onUpdate).not.toHaveBeenCalled();
      const cellGroup = screen.getAllByRole("button");
      fireEvent.click(cellGroup.at(0) as Element);

      expect(defaultProps.onUpdate).toHaveBeenCalled();
    });
  });
});
