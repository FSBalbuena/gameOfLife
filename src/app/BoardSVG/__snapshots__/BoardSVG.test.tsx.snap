// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardSVG Renders without crashing 1`] = `
<div>
  <svg
    aria-disabled="false"
    class="w-full sm:w-2/3 aspect-square border"
    role="application"
  >
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="0%"
      y="0%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="0%"
      y="20%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="0%"
      y="40%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="0%"
      y="60%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="0%"
      y="80%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="20%"
      y="0%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="20%"
      y="20%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="20%"
      y="40%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="20%"
      y="60%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="20%"
      y="80%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="40%"
      y="0%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="40%"
      y="20%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="40%"
      y="40%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="40%"
      y="60%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="40%"
      y="80%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="60%"
      y="0%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="60%"
      y="20%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="60%"
      y="40%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="60%"
      y="60%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="60%"
      y="80%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="80%"
      y="0%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="80%"
      y="20%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="80%"
      y="40%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="80%"
      y="60%"
    />
    <rect
      class="stroke-gray-400 stroke-1"
      fill="white"
      height="20%"
      role="button"
      width="20%"
      x="80%"
      y="80%"
    />
  </svg>
</div>
`;
